### ISSUE TITLE
Twisted Web WSGI container sometimes emits too many (or duplicate) headers

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3569]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3569|
|Type|defect|
|Created|2008-12-09 17:30:54Z|
|Branch|https://github.com/twisted/twisted/tree/wsgi-header-control-3569|

If a wsgi application supplies _server_, _date_, or _content-type_ headers to `start_response`, the HTTP server will emit multiple header lines.  Also, if _content-type_ is not supplied, the server will emit a _content-type_ header anyway.

_content-type_ should only be emitted if the application supplies it.  The application should not be able to override _server_ and _date_.

See [#2753](https://github.com/twisted/twisted/issues/2753) for the original report.

<details><summary>Searchable metadata</summary>

```
trac-id__3569 3569
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_wsgi_header_control_3569 branches/wsgi-header-control-3569
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1228843854000000 1228843854000000
changetime__1228928760000000 1228928760000000
version__None None
owner__ 
cc__wangchun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [25606]) Branching to 'wsgi-header-control-3569'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [25607]) add tests for some other wsgi app response header generation and fix the wsgi container behavior for those cases

refs [#3569](https://github.com/twisted/twisted/issues/3569)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

There is an extra blank line before test_applicationProvidedContentType.

+1!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [25615]) Merge wsgi-header-control-3569

Author: exarkun, itamarst
Reviewer: therve
Fixes: [#3569](https://github.com/twisted/twisted/issues/3569)

Fix header generation in the WSGI container so that it:

    * only emits a Content-Type if the application object asks it to
    * ignores any values provided for the server or date headers provided by
      the application

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
